src/wasm-binary.h

```diff
@@ -792,33 +866,651 @@ class WasmBinaryWriter : public WasmVisitor<void> {
+    return ret;
+  }
+  float getFloat32() {
+    if (debug) std::cerr << "<==" << std::endl;
+    auto ret = Literal(getInt32()).reinterpretf32();
+    if (debug) std::cerr << "getFloat32: " << ret << " ==>" << std::endl;
+    return ret;
+  }
+  double getFloat64() {
+    if (debug) std::cerr << "<==" << std::endl;
+    auto ret = Literal(getInt64()).reinterpretf64();
+    if (debug) std::cerr << "getFloat64: " << ret << " ==>" << std::endl;
+    return ret;
+  }
+
+  uint32_t getLEB128() {
+    if (debug) std::cerr << "<==" << std::endl;
+    LEB128 ret;
+    ret.read([&]() {
+      return getInt8();
+    });
+    if (debug) std::cerr << "getLEB128: " << ret.value << " ==>" << std::endl;
+    return ret.value;
+  }
+  WasmType getWasmType() {
+    int8_t type = getInt8();
+    switch (type) {
+      case 0: return none;
+      case 1: return i32;
+      case 2: return i64;
+      case 3: return f32;
+      case 4: return f64;
+      default: abort();
+    }
+  }
+
+  Name getString() {
+    if (debug) std::cerr << "<==" << std::endl;
+    size_t offset = getInt32();
+    Name ret = cashew::IString((&input[0]) + offset, false);
+    if (debug) std::cerr << "getString: " << ret << " ==>" << std::endl;
+    return ret;
+  }
+
+  void verifyInt8(int8_t x) {
+    int8_t y = getInt8();
+    assert(x == y);
+  }
+  void verifyInt16(int16_t x) {
+    int16_t y = getInt16();
+    assert(x == y);
+  }
+  void verifyInt32(int32_t x) {
+    int32_t y = getInt32();
+    assert(x == y);
+  }
+  void verifyInt64(int64_t x) {
+    int64_t y = getInt64();
+    assert(x == y);
+  }
+  void verifyFloat32(float x) {
+    float y = getFloat32();
+    assert(x == y);
+  }
+  void verifyFloat64(double x) {
+    double y = getFloat64();
+    assert(x == y);
+  }
+
+  void readMemory() {
+    if (debug) std::cerr << "== readMemory" << std::endl;
+    verifyInt8(BinaryConsts::Memory);
+    wasm.memory.initial = pow(2, getInt8());
+    wasm.memory.max = pow(2, getInt8());
+    verifyInt8(1); // export memory
+  }
+
+  void readSignatures() {
+    if (debug) std::cerr << "== readSignatures" << std::endl;
+    verifyInt8(BinaryConsts::Signatures);
+    size_t numTypes = getLEB128();
+    if (debug) std::cerr << "num: " << numTypes << std::endl;
+    for (size_t i = 0; i < numTypes; i++) {
+      if (debug) std::cerr << "read one" << std::endl;
+      auto curr = allocator.alloc<FunctionType>();
+      size_t numParams = getInt8();
+      if (debug) std::cerr << "num params: " << numParams << std::endl;
+      curr->result = getWasmType();
+      for (size_t j = 0; j < numParams; j++) {
+        curr->params.push_back(getWasmType());
+      }
+      wasm.addFunctionType(curr);
+    }
+  }
+
+  std::vector<Name> mappedFunctions; // index => name of the Import or Function
+
+  size_t nextLabel;
+
+  Name getNextLabel() {
+    return cashew::IString(("label$" + std::to_string(nextLabel++)).c_str(), false);
```

Comment:
perhaps worth doing as a general refactoring for IString. But it's used in a bunch of places, so it needs more consideration. followup discussion seems best.